examples/checkpoint/source.py

```diff
@@ -0,0 +1,98 @@
+import time
+from copy import copy
+from collections import deque
+from erdos.data_stream import DataStream
+from erdos.message import Message, WatermarkMessage
+from erdos.op import Op
+from erdos.timestamp import Timestamp
+from erdos.utils import setup_logging
+import checkpoint_util
+
+
+class Source(Op):
+    def __init__(self,
+                 name,
+                 checkpoint_freq=10,
+                 state_size=10,
+                 num_messages=50,
+                 fps=10,
+                 log_file_name=None):
+        super(Source, self).__init__(name)
+        self._logger = setup_logging(self.name, log_file_name)
+        self._seq_num = 1
+        self._num_messages = num_messages
+        self._time_gap = 1.0 / fps
+
+        self._state_size = state_size
+        self._state = deque()
+        self._checkpoint_freq = checkpoint_freq
+        self._checkpoints = dict()
+
+
+    @staticmethod
+    def setup_streams(input_streams):
+        input_streams\
+            .filter(checkpoint_util.is_control_stream)\
+            .add_callback(Source.on_rollback_msg)
+        return [DataStream(name='input_stream')]
+
+    def on_rollback_msg(self, msg):
```

Comment:
There's a lot of repetition in these on_rollback_msg methods. Let's discuss in person to see if we can figure a good way to reduce the repetition.